tools/chpl-language-server/src/chpl-language-server.py

```diff
@@ -0,0 +1,331 @@
+            ResolvedPair(NodeAndRange(node), NodeAndRange(to))
+            for node, _ in chapel.each_matching(asts, chapel.core.Identifier)
+            if (to := node.to_node())
+        ]
+        self.segments.sort(key=lambda s: s.ident.rng.start)
+        self.siblings = chapel.SiblingMap(self.get_asts())
+
+
+def run_lsp():
+    """
+    Start a language server on the standard input/output, and use it to
+    report linter warnings as LSP diagnostics.
+    """
+
+    server = LanguageServer("chplcheck", "v0.1")
+
+    contexts: Dict[str, FileInfo] = {}
+
+    def get_context(uri: str, do_update: bool = False) -> Tuple[FileInfo, List[Any]]:
+        """
+        The LSP driver maintains one Chapel context per-file. This is to avoid
+        having to reset all files' text etc. when a single file is updated.
+        There may be a more principled approach we can take in the future.
+
+        This function returns an _update_ context, which is effectively a context
+        in which we can save / make use of updated file text. If there wasn't
+        a context for a URI, a brand new context will do. For existing contexts,
+        an older version of the file's text is probably stored, so advance
+        the context to next revision, invalidating that cache.
+
+        Thus, this method is effectively allocate-or-advance-context.
+        """
+
+        errors = []
+
+        if uri in contexts:
+            file_info = contexts[uri]
+            if do_update:
+                file_info.context.advance_to_next_revision(False)
+                with file_info.context.track_errors() as errors:
+                    file_info.rebuild_segments()
+        else:
+            context = chapel.core.Context()
+            with context.track_errors() as errors:
+                file_info = FileInfo(uri, context)
+            contexts[uri] = file_info
+
+        return (file_info, errors)
+
+    def build_diagnostics(uri):
+        """
+        Parse a file at a particular URI, run the linter rules on the resulting
+        ASTs, and return them as LSP diagnostics.
+        """
+
+        fi, errors = get_context(uri, do_update=True)
+        with fi.context.track_errors() as new_errors:
+            _ = fi.parse_file()
+        errors.extend(new_errors)
+
+        diagnostics = []
+
+        kind_to_severity = {
+            "error": DiagnosticSeverity.Error,
+            "syntax": DiagnosticSeverity.Error,
+            "note": DiagnosticSeverity.Information,
+            "warning": DiagnosticSeverity.Warning,
+        }
+
+        for error in errors:
+            diagnostic = Diagnostic(
+                range=location_to_range(error.location()),
+                message="{}: [{}]: {}".format(
+                    error.kind().capitalize(), error.type(), error.message()
+                ),
+                severity=kind_to_severity[error.kind()],
+            )
+            diagnostics.append(diagnostic)
+        return diagnostics
```

Comment:
Not necessarily in this PR, but this is another good candidate for moving to `chapel.lsp`.

Reply:
I agree, I think thats easy to do now